Simbody/src/CableSpan.cpp

```diff
@@ -0,0 +1,2354 @@
+{
+    assert(isInstanceOf(s));
+    return static_cast<CableSubsystem&>(s);
+}
+
+CableSubsystem::CableSubsystem()
+{
+    adoptSubsystemGuts(new Impl());
+}
+
+CableSubsystem::CableSubsystem(MultibodySystem& mbs)
+{
+    adoptSubsystemGuts(new Impl());
+    mbs.adoptSubsystem(*this);
+}
+
+int CableSubsystem::getNumCables() const
+{
+    return getImpl().getNumCables();
+}
+
+const CableSpan& CableSubsystem::getCable(CableSpanIndex ix) const
+{
+    return getImpl().getCable(ix);
+}
+
+CableSpan& CableSubsystem::updCable(CableSpanIndex ix)
+{
+    return updImpl().updCable(ix);
+}
+
+//------------------------------------------------------------------------------
+//                        CABLE SUBSYSTEM IMPL
+//------------------------------------------------------------------------------
+
+const CableSubsystem::Impl& CableSubsystem::getImpl() const
+{
+    return SimTK_DYNAMIC_CAST_DEBUG<const Impl&>(getSubsystemGuts());
+}
+
+CableSubsystem::Impl& CableSubsystem::updImpl()
+{
+    return SimTK_DYNAMIC_CAST_DEBUG<Impl&>(updSubsystemGuts());
+}
+
+int CableSubsystem::Impl::calcDecorativeGeometryAndAppendImpl(
+    const State& state,
+    Stage stage,
+    Array_<DecorativeGeometry>& decorations) const
+{
+    if (stage != Stage::Position)
+        return 0;
+
+    for (const CableSpan& cable : cables) {
+        int returnValue = cable.getImpl().calcDecorativeGeometryAndAppend(
+            state,
+            stage,
+            decorations);
+        if (returnValue != 0) {
+            return returnValue;
+        }
+    }
+    return 0;
+}
+
+int CableSubsystem::Impl::realizeSubsystemTopologyImpl(State& state) const
+{
+    // Briefly allow writing into the Topology cache; after this the
+    // Topology cache is const.
+    Impl* wThis = const_cast<Impl*>(this);
+
+    wThis->realizeTopology(state);
+    for (CableSpanIndex ix(0); ix < cables.size(); ++ix) {
+        CableSpan& path = wThis->updCable(ix);
+        path.updImpl().realizeTopology(state);
+    }
+
+    return 0;
+}
+
+//==============================================================================
+//                                CABLE SPAN
+//==============================================================================
+
+// Allocate a new default-constructed CableSpan::Impl.
+CableSpan::CableSpan() : m_Impl(std::shared_ptr<Impl>(new Impl())) {}
+
+// Invalidate the CableSpan::Impl, and delete the reference. Anyone still
+// holding on to the data (like the subsystem) will know it is invalid, and
+// delete their reference as well.
+CableSpan::~CableSpan()
+{
+    if (m_Impl) {
+        m_Impl->invalidateSubsystemEntry();
+        m_Impl = nullptr;
+    }
+}
+
+CableSpan::CableSpan(const CableSpan& source) :
+    m_Impl(
+        source.m_Impl ? std::shared_ptr<Impl>(new Impl(source.getImpl()))
+                      : nullptr)
+{}
+
+CableSpan& CableSpan::operator=(const CableSpan& source)
+{
+    return *this = CableSpan(source);
```

Comment:
All of this code is giving me the impression that `std::shared_ptr` is the wrong call. If you don't want shared ownership (and your copy ctor and copy assignment operation indicate this by copy-constructing a new `Impl`) then you should use something like `SimTK::ClonePtr` or similar instead

Reply:
The behavior now matches that of a `MobilizedBody`: The `CableSpan` acts as a handle to the data/implementation `CableSpan::Impl`. A copy is cheap (increases ref count). The Subsystem will keep the data alive if your `CableSpan` goes out of scope. So:

```cpp
CableSubsystem cables(system);

// Register new cable:
{
    CableSpan cable(cables, ...);
    cables.getNumCables(); // = 1
}

// CableSubsystem still has a reference.
cables.getNumCables(); // = 1
```

I came to the conclusion that the ownership is shared between the subsystem and the owner of the handle. Not sure if the intention is to distribute these handles, but it looks like you can do that.